Source/Core/Common/CircularQueue.h

```diff
@@ -0,0 +1,258 @@
+  // Elements may be written here before using AdvanceWriteHead.
+  std::span<ElementType> GetWriteSpan()
+  {
+    return std::span(&GetWriteHead(), m_storage.size() - m_write_index);
+  }
+
+  // Push elements from range.
+  // Will overwrite existing data if Capacity is exceeded.
+  template <std::ranges::range R>
+  requires(std::is_convertible_v<typename R::value_type, ElementType>)
+  void PushRange(const R& range)
+  {
+    auto input_iter = range.begin();
+    const std::size_t range_size = std::size(range);
+    std::size_t remaining = range_size;
+    auto output_iter = m_storage.begin() + m_write_index;
+
+    while (remaining != 0)
+    {
+      const std::size_t count = std::min<std::size_t>(m_storage.end() - output_iter, remaining);
+      input_iter = std::ranges::copy_n(input_iter, count, output_iter).in;
+
+      remaining -= count;
+      output_iter = m_storage.begin();
+    }
+
+    AdvanceWriteHead(range_size);
+  }
+
+  // May be used by the "producer" thread.
+  // Moves the write position and increases the size without writing any data.
+  void AdvanceWriteHead(std::size_t count, std::memory_order order = std::memory_order_release)
+  {
+    m_write_index = (m_write_index + count) % Capacity();
+    AdjustSize(count, order);
+  }
+
+  // "Pop" functions are only safe from the "consumer" thread.
+  const ElementType& GetReadHead() const { return m_storage[m_read_index]; }
+
+  ElementType Pop()
+  {
+    assert(Size() != 0);
+
+    ElementType result = GetReadHead();
+    AdvanceReadHead(1);
+    return result;
+  }
+
+  // Returns a span from the read-head to the end of the buffer.
+  // Not Size() checked. Caller must verify elements have been written.
+  std::span<const ElementType> GetReadSpan() const
+  {
+    return std::span(&GetReadHead(), m_storage.size() - m_read_index);
+  }
+
+  // Pop elements to fill the provided range.
+  // Does not check first if this many elements are readable.
+  template <std::ranges::range R>
+  requires(std::is_convertible_v<ElementType, typename R::value_type>)
+  void PopRange(R* range)
+  {
+    auto output_iter = range->begin();
+    const std::size_t range_size = std::size(*range);
+    std::size_t remaining = range_size;
+    auto input_iter = m_storage.begin() + m_read_index;
+
+    while (remaining != 0)
+    {
+      const std::size_t count = std::min<std::size_t>(m_storage.end() - input_iter, remaining);
+      output_iter = std::ranges::copy_n(input_iter, count, output_iter).out;
+
+      remaining -= count;
+      input_iter = m_storage.begin();
+    }
+
+    AdvanceReadHead(range_size);
+  }
+
+  // Used by the "consumer" thread to move the read position without reading any data.
+  // Also reduces size by the given amount.
+  // Use RewindReadHead for negative values.
+  void AdvanceReadHead(std::size_t count)
+  {
+    m_read_index = (m_read_index + count) % Capacity();
+    AdjustSize(0 - count, std::memory_order_relaxed);
```

Comment:
`0 - count` underflows here, which is cancelled out by `m_size += count` (or fetch_add) overflowing in `AdjustSize`. The math works, but I feel like it'd be cleaner to have a bit of code duplication with separate `IncreaseSize` and `DecreaseSize` functions.

Reply:
I dunno, that's 10 lines of duplicate code just for being afraid of some unsigned integer underflow. :D

Reply:
Unsigned underflow is well defined, so I'd say take advantage of it.